src/v/cluster/partition.h

```diff
@@ -28,7 +29,6 @@
 
 namespace cluster {
 class partition_manager;
-class archival_metadata_stm;
```

Comment:
>making it fwd declaration
means that anyone including the header has to make sure to include other
headers.

I'm kind of confused about this. Do you mean callers that want to use `partition::archival_meta_stm()` need to include archival_metadata_stm.h? Or that just including `cluster/partition.h` always mandates also including `archival/archival_metadata_stm.h`? If the former, that seems like expected behavior? And if the latter, it's not quite clear to me why that is -- could you point out why that is?

Reply:
What I mean is that `archival_metadata_stm` is part of the interface, rather than the implementation. So if my module includes `partition.h` then only part of the public interface exported by the header will work without including additional headers. the fwd declaration is usually only used when the interface takes a reference and the definition is imported in the .cc file. example:

```cpp
class {
public:
  archival_metadata_stm& get_stm();
};
```

vs

```cpp
class {
private:
  archival_metadata_stm& stm;
};
```

thoughts?

Reply:
discussed offline. generally it's nice to have headers not require other headers to be able to use their interfaces. this particular change was just to get something to compile i suppose i could have / should have had the user include the archival stm header instead. alternatively, and i think preferred, is the more effort required approach of making the archival stm header cheaper to include.

Reply:
definitely feel free to change back. the module conversion project is going to be 100 steps forward and 20 steps backwards. a second pass after conversion can resolve all of the remaining header-only dependencies and try to shrink the headers that are published publicaly.